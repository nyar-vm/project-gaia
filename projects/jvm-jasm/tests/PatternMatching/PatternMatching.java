public class PatternMatching {
    
    public String instanceofPattern(Object obj) {
        if (obj instanceof String s) {
            return "String with length: " + s.length();
        }
        return "Not a string";
    }
    
    public String switchExpression(Object obj) {
        return switch (obj) {
            case String s when s.length() > 5 -> "Long string: " + s;
            case String s -> "Short string: " + s;
            case Integer i -> "Number: " + i;
            case null -> "Null value";
            default -> "Unknown type";
        };
    }
    
    public int recordPattern(Object obj) {
        record Point(int x, int y) {}
        record Line(Point start, Point end) {}
        
        if (obj instanceof Line(Point(int x1, int y1), Point(int x2, int y2))) {
            return Math.abs(x2 - x1) + Math.abs(y2 - y1);
        }
        return 0;
    }
    
    public String sealedClassPattern(Shape shape) {
        return switch (shape) {
            case Circle c -> "Circle with radius " + c.radius();
            case Rectangle r -> "Rectangle " + r.width() + "x" + r.height();
        };
    }
    
    sealed interface Shape permits Circle, Rectangle {}
    record Circle(double radius) implements Shape {}
    record Rectangle(double width, double height) implements Shape {}
}