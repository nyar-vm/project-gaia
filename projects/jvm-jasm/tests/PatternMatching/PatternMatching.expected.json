{
    "class_name": "PatternMatching",
    "class_modifiers": [
        "public",
        "super"
    ],
    "method_count": 2,
    "field_count": 0,
    "expected_methods": [
        "<init>:()V",
        "instanceofPattern:(Ljava/lang/Object;)Ljava/lang/String;"
    ],
    "expected_fields": [],
    "syntactic_sugar_features": [],
    "check_tokens": true,
    "source_file": null
}